### ISSUE TITLE
Fingerprint reader support [$65]

### ISSUE THREAD

**Issue by userosos** [keepassxreboot/keepassxc#3337]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2490%20Funded-%2300A156.svg" />](https://issuehunt.io/r/keepassxreboot/keepassxc/issues/3337)

Hi! Can you add support Fingerprint? My laptop has fingerprint reader i want use it for unblock database.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $90.00)

- [<img src='https://avatars.githubusercontent.com/u/43879831?v=4' alt='kbarnawi' width=24 height=24> kbarnawi](https://issuehunt.io/u/kbarnawi) ($5.00)
- [<img src='https://avatars.githubusercontent.com/u/45434519?v=4' alt='githubuser325' width=24 height=24> githubuser325](https://issuehunt.io/u/githubuser325) ($20.00)
- $65.00 have been anonymously funded.

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/keepassxreboot/keepassxc/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by droidmonkey** [comment]
It is certainly possible but biometric sensors (especially on laptops) are notoriously terrible. They usually are meant to interoperate with proprietary software written by the laptop vendor and don't offer any functionality beyond that. Windows does have a robust biometrics interface, but that requires the hardware vendor to actually write a driver that works with it.

https://docs.microsoft.com/en-us/windows/desktop/secbiomet/creating-client-applications#manage-credentials

Over in Linux land your mileage can vary even further. MacOS has TouchID which we have already integrated.

**Comment by phoerious** [comment]
On Windows we can use Windows Hello.

**Comment by userosos** [comment]
@phoerious 
> On Windows we can use Windows Hello.
Can you make howto?

**Comment by droidmonkey** [comment]
Windows hello is the manifestation of the biometrics service. We don't support that right now.

**Comment by userosos** [comment]
> Windows hello is the manifestation of the biometrics service. We don't support that right now.

Ok. i understand. Do you have an planned for support it?

**Comment by droidmonkey** [comment]
Not anytime soon

**Comment by goetzc** [comment]
On GNU/Linux this would be done around _fprint_ which seems to be the thing.

More info:
https://www.freedesktop.org/wiki/Software/fprint/
http://www.linux-pam.org/

**Comment by HyperionPhilipsHue** [comment]
Please add Support for Windows Hello soon! It's the only missing thing in your beautiful Software <3

**Comment by MisterY** [comment]
Windows-specific support (Hello) is here - #2462.

**Comment by GitHubUser325** [comment]
> "droidmonkey changed the title Fingerprint reader support Fingerprint reader support [$20]"

Since I'm not so common with github, what means the [$20]? I am ready to contribute sth, if you can implement the Windows Hello Support :)

**Comment by droidmonkey** [comment]
It means there is a bounty on this issue of $20. I was going to add a more substantial bounty to the windows hello and this issue.

**Comment by andreas512** [comment]
I was also interested in a fingerprint support for keepassxc. How can I contribute to the bounty, in order to push the topic a bit?

**Comment by goetzc** [comment]
@andreas512 this is the link:  
https://www.bountysource.com/issues/76210442-fingerprint-reader-support

@droidmonkey is it possible to integrate BoutySource with GH so that it shows or updates here the current bounty status?

**Comment by droidmonkey** [comment]
I am not using bounty source anymore, their platform is broken and they are not investing in it.

**Comment by goetzc** [comment]
In fact I was getting 500 errors before, it looks like a bit abandoned.

What will happen with all the issue-specific donations?

**Comment by droidmonkey** [comment]
My assumption is they are intact until the site goes away forever. I am no longer sending money through them. We moved to issuehunt.io.

**Comment by goetzc** [comment]
Good to know, then the link for this issue is: 
https://www.bountysource.com/issues/76210442-fingerprint-reader-support

**Comment by GitHubUser325** [comment]
Would the Contribution be only for the Fingerprint Function, or generally for Windows Hello? I ask, since I use an IR camera with Windows Hello, only Fingerprint Sensor would be of little use..

**Comment by goetzc** [comment]
@GitHubUser325 for Windows Hello the GitHub issue is https://github.com/keepassxreboot/keepassxc/issues/2462 and the IssueHunt bonuty is [this](https://issuehunt.io/r/keepassxreboot/keepassxc/issues/2462).

**Comment by jduo** [comment]
@GitHubUser325 are you still working on this issue? I'd be interested in taking it on.

**Comment by andreas512** [comment]
@jduo: I don't get the impression from the discussion that GitHubUser325 is working on this issue! I think he is also seeking a solution...
@goetzc: I understand that #2462 is dedicated to a fingerprint solution by Windows Hello - but it is not funded yet.
But which solution is #3337 dedicated to? Would Windows Hello also be an option for this topic here? (already funded!)

I am looking for a solution under Win10 on a HP notebook, so Windows Hello would work fine for me - but I am open for other solutions as well. And I am willing to contribute to the funding. 
But which issue is the right one to fund?

**Comment by StefanKrapf** [comment]
Hi
I just contributed this topic with $15 to fill the gap to $70.
Sadly the other topic is closed now (https://github.com/keepassxreboot/keepassxc/issues/2462), but maybe I have a little piece to a solution for a C++ Library: https://community.bitwarden.com/t/fingerprint-support-all-platforms-to-do-not-re-prompt-the-master-password-all-the-time/143/48
But I really don't know because I'm not a programmer!
But hopefully it can help to fulfill the dream of the Windows Hello solution with fingerprint scanner.
Thanks a lot for your work
Stefan

PS: At the moment, this feature still prevents me from switching from KeePass to KeePassXC!

**Comment by smlu** [comment]
@andreas512 have you found any solution for Windows Hello support or did you maybe do any work on it? 
Btw, I'm currently exploring how Windows Hello could be used to provide bio-metrics based securing of master key on Windows.

**Comment by phoerious** [comment]
Not yet, but we are in the process of migrating to the Windows toolchain, so that may open up opportunities to implement Hello.

**Comment by smlu** [comment]
@phoerious, I managed to get Window Hello working directly via Win32 API, using Cryptography API and Windows Credential API.
In short, the persistent symmetric encryption key is generated and stored by Microsoft Passport storage provider, this key is used to encrypt KeePassXC master password. The encrypted key is then stored in windows credential manager. To decrypt encrypted key user must authenticate via Windows Hello and decrypted key is then passed back to the app.

I have the implementation almost ready, just need to handle few more things. But if desired, I can make PR now and push updates later before it's merged.

This is how it looks in action:
![whkpxc3](https://user-images.githubusercontent.com/4377556/105561583-62e8ff80-5d17-11eb-983a-a4ceb79906f1.gif)

**Comment by droidmonkey** [comment]
Thats amazing

**Comment by goetzc** [comment]
@droidmonkey what about the Linux support? As there is a Windows-specific issue https://github.com/keepassxreboot/keepassxc/issues/2462 for it, with the IssueHunt https://issuehunt.io/r/keepassxreboot/keepassxc/issues/2462

My point is that I contributed to the bounties for Linux support, and have posted comments on this thread mentioning Linux, I'm personally not interested in Windows.

**Comment by droidmonkey** [comment]
Thank you for your contributions. I doubt greatly anyone is going to implement both Windows and Linux to gain this bounty. Linux is far harder since there is no guarantee of performance and no crypto store like what windows hello provides. There is no bounty on the windows hello issue, thus this bounty applies since this is the generic issue.

**Comment by phoerious** [comment]
@smlu That's great. Submit the PR and add your updates over time. That's always better than submitting the finished thing in one big (or bug) update, since we can have a look at it early.

As for @goetzc's concerns, I would suggest you build a generic API around it that also abstracts away macOS TouchID, which is already supported. We can then replace the -DWITH_XC_TOUCHID compile flag with a more generic -DWITH_XC_FINGERPRINT flag and contribute a libfprint implementation for Linux later. It's a freedesktop project that aims to support most readers on Linux and it looks actively maintained, so we can give it a shot.

**Comment by smlu** [comment]
@phoerious @droidmonkey I'll make PR and we can discus further details over Windows Hello implementation there.

As for bounty goes, I don't want it :). I've just recently discovered this awesome project and noticed that windows hello support is missing. I intended  to write support for it no matter if there is a bounty or not. (I just wrote in this thread because #2462 is locked and I didn't want open new issue or preliminary make PR). 
Therefore, I donate any bounty to you guys!

**Comment by goetzc** [comment]
Created a [Linux-specific](https://github.com/keepassxreboot/keepassxc/issues/5991) ticket, to avoid any further confusion with the Windows ticket 😄 

Might be worth adding both ticket links to the first comment on this same issue?

**Comment by Yoann166** [comment]
> @phoerious, I managed to get Window Hello working directly via Win32 API, using Cryptography API and Windows Credential API.
> In short, the persistent symmetric encryption key is generated and stored by Microsoft Passport storage provider, this key is used to encrypt KeePassXC master password. The encrypted key is then stored in windows credential manager. To decrypt encrypted key user must authenticate via Windows Hello and decrypted key is then passed back to the app.
> 
> I have the implementation almost ready, just need to handle few more things. But if desired, I can make PR now and push updates later before it's merged.
> 
> This is how it looks in action:
> ![whkpxc3](https://user-images.githubusercontent.com/4377556/105561583-62e8ff80-5d17-11eb-983a-a4ceb79906f1.gif)

Wow! Great! Can't wait for this to be available!

**Comment by smlu** [comment]
@Yoann166, if you're able to use Visual Studio and compile the code yourself, you can clone my [win-dev](https://github.com/smlu/keepassxc/tree/win-dev) branch and try it out.

**Comment by Plinsboorg** [comment]
Is someone working on it? Looks like there is some solution that needs to be checked.
I'm also waiting for this feature.

**Comment by droidmonkey** [comment]
It's been worked, we have other things to work on first before this becomes a reality.

**Comment by QuAzI** [comment]
Any plans to release it?

**Comment by GitHubUser325** [comment]
Is there a new status info? I am so eagerly waiting for it :(

**Comment by michaelk83** [comment]
> Is there a new status info?

1. MacOS TouchID is [already suported](https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-506916332).
2. ~~Need to finish the [Visual Studio build support (#5874)](https://github.com/keepassxreboot/keepassxc/pull/5874). It was waiting on the Botan refactor, which is now merged. @smlu updated the PR in Apr, so it's awaiting review.~~ done
3. ~~After that, [Windows Hello support (#6029)](https://github.com/keepassxreboot/keepassxc/pull/6029) can be continued.~~ replaced by #7384 .
4. Combining MacOS TouchID and Windows Hello code, [a common interface needs to be implemented](https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-765887978) (may be part of ~~#6029~~ #7384 ?)
5. Finally, [Linux support via fprint (#5991)](https://github.com/keepassxreboot/keepassxc/issues/5991) can be implemented on top of that common interface.

**Comment by michaelk83** [comment]
@smlu Could you maybe pull out the common interface to its own PR without the Visual Studio dependency? Then it can be merged more quickly, and someone on Linux can try tackling that side without waiting for Windows Hello.

**Comment by smlu** [comment]
The #6029 draft PR should probably work also with latest mingw that supports C++17 standard or newer. The suggested VS PR requirement is due to the fact the WindowsHello patch was built with VisualStudio (as probably every new KPXC Windows release should be :) and was not tested with other compilers.

 There are still few things that should be considered before #6029 is merged:
 * KPXC database migration procedure. 
 When master key for the DB is stored in windows credentials (WindowsHello encrypts the master key and encrypted key is stored on the system via windows credential API), it's mapped under the absolute database file path. If user moves the DB file to another location the master key is preserved in the windows credential storage but the KPXC looses the knowledge of it. Windows credential storage API might help mitigate this problem, specifically [CredEnumerate](https://docs.microsoft.com/en-us/windows/win32/api/wincred/nf-wincred-credenumeratea), but I haven't tested so I can't comment.
 *  Implement storing of master keys for 1Password vault and KeePass 1 DB
 *  Implement resting of WindowsHello storage for stored KPXC master keys from command line. 
 When KPXC is uninstalled from the user's Windows machine all stored (encrypted) master keys should also be deleted from the windows credential storage. This could be achieved through uninstaller by calling reset command on KPXC or kpxc-cli.
 * Common interface for system key storage (WindowsHello, Touch ID, Linux...). 
 This could be pushed to another PR on top of #6029.
 
 
 Unfortunately, I was very busy these few months and I was not able to continue the work on PR #6029 but I'm planning to resume. I was also hoping for some sort of review or suggestion comments on #6029  before I continue the work.

**Comment by michaelk83** [comment]
What I'm suggesting is that the Linux side doesn't need the whole of #6029, which can still take a while. It only needs the common bits:
> Common interface for system key storage (WindowsHello, Touch ID, Linux...).

Aka step 4 in my [above TODO list](https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-854550917).  
If that can be extracted and merged separately, then the Linux side can progress in parallel (if someone picks it up).

> This could be pushed to another PR on top of #6029.

The idea is to flip that around: pull that to its own PR *without* the dependencies, merge it, then have #6029 depend on that (+ on Visual Studio, as currently).

**Comment by smlu** [comment]
I agree, the common system key storage interface should probably be done in another PR and the rest built around it. 
Note though, the #6029 doesn't contain any such interface at the moment. The WindowHello related code is intertwined in preproc macros in the UI code base.

I'm also willing to make the PR for such interface (if nobody else will), but I'll need some more input from the community and maintainers before I begin.

**Comment by michaelk83** [comment]
Since it's more or less needed for Windows Hello, I think you're the primary candidate atm. :)

There are two pieces that can potentially be abstracted:
1. The credential (encrypted DB key) storage, which is also needed for regular QuickUnlock (#488).
2. The fingerprint reader interface, if that's handled directly by KPXC.

Your [suggestion in the PR](https://github.com/keepassxreboot/keepassxc/pull/6029#issuecomment-770297945) looks like a good start for (1).

(2) may be more difficult, but if either your Windows Hello code or the existing TouchID code doesn't actually handle the reader directly, then it's not needed at this stage. It can be abstracted when the Linux code is implemented.

**Comment by smlu** [comment]
> Since it's more or less needed for Windows Hello, I think you're the primary candidate atm. :)


> There are two pieces that can potentially be abstracted:
> 
>     1. The credential (encrypted DB key) storage, which is also needed for regular QuickUnlock (#488).
> 
>     2. The fingerprint reader interface, if that's handled directly by KPXC.

If the fingerprint scanning is only needed for the purpose of authenticating user before accessing the system encrypted secret keys I'd merged the fp reader interface into 1.  Especially due to the fact that the extra fp reader interface will probably be needed only for Linux platforms (your last point). In Windows case, the Windows Passport API abstract away user authentication, I'm guessing so does TouchID.

In my opinion the system credential storage interface should abstract away any required authentication and just expose simple API to the KPXC code base. It's less flexible, but more robust and gets you cleaner main code base. 
The implementation should then take care of all required steps to retrieve the secret key.

**Comment by droidmonkey** [comment]
@smlu is 100% correct. There is no user interface needed for windows or Mac, the notice to touch the reader is provided by the operating system.

**Comment by michaelk83** [comment]
> if either your Windows Hello code or the existing TouchID code doesn't actually handle the reader directly, then it's not needed at this stage.

On 2nd thought, the TouchID and Windows Hello themselves should (ideally) be abstracted, so that they become OS-specific backends. Then the fprint code will be the Linux backend for the same.

But, on 3rd thought, I suppose this all *could* be handled with just OS-specific `ifdef`s for now, the same way TouchID was implemented, and leave the abstraction refactor for later. It would just be uglier that way.

> If the fingerprint scanning is only needed for the purpose of authenticating user before accessing the system encrypted secret keys I'd merged the fp reader interface into 1.

Yes, maybe. But keep in mind that for regular QuickUnlock we'll need the credentials storage without the biometrics. In that case, the DB passphrase is encrypted not by the fingerprint, but by something else (probably by the last few chars of the passphrase, or their hash; during unlock, the user will enter those chars manually). So at least the storage and retrieval of the encrypted passphrase should be independent from the biometrics.

From your PR description:

> CNG creates RSA 2048 bit encryption key via Microsoft Passport storage provider and "locks" it on Windows Hello authenticator when key is used for decryption. ... This key is then used to encrypt database key, and encrypted database key is stored on windows using Credential API.

So these two parts should be separate, so that instead of the Microsoft Passport encryption key, regular QuickUnlock can use whatever arbitrary encryption key.

**Comment by michaelk83** [comment]
> There is no user interface needed for windows or Mac, the notice to touch the reader is provided by the operating system.

To clarify, I was talking about code interfaces, not user interfaces. That said, there is *some* UI for working with TouchID, e.g. in the `DatabaseOpenWidget` and settings, which will likely be rather similar (and mutually-exclusive) for Windows Hello and Linux fprint. There may be room for generalization there. I guess that's what `OSKeychain::name()` will be for.

---

I took a look at the Touch ID code. First thing I noticed is that the [`TouchID` class interface](https://github.com/keepassxreboot/keepassxc/blob/develop/src/touchid/TouchID.h) is very similar to @smlu 's [`OSKeychain` proposal](https://github.com/keepassxreboot/keepassxc/pull/6029#issuecomment-770297945). Probably not a coincidence?

The 2nd thing I learned is that the TouchID code [does things "backwards"](https://github.com/keepassxreboot/keepassxc/blob/develop/src/touchid/TouchID.mm#L38). It stores *the encryption key* in MacOS keychain (protected by the TouchID), while the encrypted passphrase is kept in regular memory. The encryption key is randomly generated. AFAICT, this can be changed to store *the encrypted passphrase* in the keychain (~~or both passphrase and encryption key~~) by changing the [`valueData`](https://github.com/keepassxreboot/keepassxc/blob/develop/src/touchid/TouchID.mm#L111). And probably the keychain data doesn't *have to* be protected by TouchID (in the case of regular QuickUnlock), by changing the [access control flags](https://github.com/keepassxreboot/keepassxc/blob/develop/src/touchid/TouchID.mm#L97). But both of these changes are for another issue (edit: #6644).

So I suggest the following changes to `OSKeychain`:
1. Add separate `encrypt()` and `decrypt()` methods to encrypt and decrypt the database passphrase. With Windows Hello, this would implement the Microsoft Passport part ("creates RSA 2048 bit encryption key ...").
2. Rename `storeKey()`, `getKey()`, and `removeKey()` to more generic `set(path, value)`, `get(path)`, `delete(path)`, which will work with generic key-value pairs. With Windows Hello, this would implement the Credential API part.
3. `set()` should also accept flags or an enum to specify whether the data should be protected by biometrics (access control).

Then you'd usually call `encrypt()` to encrypt the passphrase, and then pass the encrypted passphrase to the `set()` method. On the way back, you'd take the return value from `get()` and pass it to `decrypt()`.

The TouchID implementation ~~can append the encryption key and IV to the encrypted data returned from `encrypt()`, and~~ will set the access control according to the flags that are passed to `set()`. And with regular QuickUnlock, the `encrypt()` call is replaced by other encryption logic, and the flags are set to *not* use biometrics.

(edit: not quite that simple... I'll update tomorrow.)

So the class definition will be something like this:
```c++
class OSKeychain
{
    static bool isAvailable();
    static bool isPersistentStorageAvailable();
    static const QString& name();

    bool encrypt(QByteArray& value) const;
    bool decrypt(QByteArray& value) const;

    bool set(QStringView path, const QByteArray& value, const Something flags);
    QSharedPointer<QByteArray> get(QStringView path) const;
    void delete(QStringView path);
    bool reset() const;
};
```

**Comment by michaelk83** [comment]
Btw:
> This key is then used to encrypt database key, and encrypted database key is stored on windows using Credential API.

Why not use  Microsoft Passport storage to store the encrypted database passphrase as well (if biometrics is enabled)?

**Comment by michaelk83** [comment]
> The TouchID implementation can append the encryption key and IV to the encrypted data returned from `encrypt()`.

That was a Bad Idea, since we don't want to save the encryption key and encrypted data in the same place.

So there will be an issue with the TouchID implementation. Its encryption step generates a random encryption key and IV, and those need to be saved somewhere. So the `encrypt()` method, as proposed, isn't good enough.

Taking a step back, the requirements for `OSKeychain` are as follows:
1. The keychain can be used with or without biometrics.
2. Ability to use custom encryption for regular QuickUnlock (derived encryption key that isn't saved).
3. Some way to handle the encryption key and IV when they are generated by the keychain implementation (TouchID case).

(1) is handled by the flags passed to `set()`.

(2) can also be handled with a flag (encrypt/don't), but can actually be relaxed, because any internal encryption is transparent to the caller anyway. Whether the keychain implementation adds its own layer of encryption, and whether the caller passes an unencrypted passphrase or one already encrypted with a derived key, it can still all work the same.

Then, (3) becomes an internal implementation detail.

With fingerprints, the internal encryption key can be derived from the fingerprint data (and not saved), but that's also an implementation detail that doesn't affect the interface.

---

With that in mind, the class definition becomes:
```c++
class OSKeychain
{
    static bool isAvailable();
    static bool isPersistent();
    static const QString& name();

    bool clear();
    bool contains(QStringView path) const;
    bool delete(QStringView path);
    QSharedPointer<QByteArray> get(QStringView path) const;
    bool set(QStringView path, const QByteArray& value, const OSKeychain::Flags flags);
};
```
(Removed the `encrypt()` and `decrypt()` methods, added `contains()`, various cleanup.)

The flags should include at least `OSKeychain::USE_BIOMETRICS`, and maybe `OSKeychain::ENCRYPT`. The TouchID implementation can remain mostly the same as now, except for taking the flags into account. For regular QuickUnlock, the flags are cleared, and the passphrase is encrypted with a derived key before it's passed to `set()`.

**Comment by smlu** [comment]
>So these two parts should be separate, so that instead of the Microsoft Passport encryption key, regular QuickUnlock can use whatever arbitrary encryption key.

Windows Hello can't decouple encryption from user authentication (e.g. bio-metrics). When you create Microsoft Passport persistent asymmetric key (used for encrypting KPXC secrets) you also specify if user authentication is mandatory when using the key. This authn property is "burned" to the key and can't be changed.  The system then takes care of what kind of authentication is provided to the user depending on what he/she set up in Windows Hello.

Note: For Windows Hello quick unlock feature, this persistent key also has to be created (and deleted when KPXC  is uninstalled).
There is no option to get temporary key from Microsoft Passport. This is though possible in .NET land.

> I took a look at the Touch ID code. First thing I noticed is that the TouchID class interface is very similar to @smlu 's OSKeychain proposal. Probably not a coincidence?

No coincidence yes, I followed the current interface of Touch ID so there won't be needed much to refactor.

> Why not use Microsoft Passport storage to store the encrypted database passphrase as well (if biometrics is enabled)?

The Microsoft Passport API can work only with asymmetric RSA keys. Some info on this: [https://blogs.windows.com/windowsdeveloper/2016/01/26/convenient-two-factor-authentication-with-microsoft-passport-and-windows-hello/](https://blogs.windows.com/windowsdeveloper/2016/01/26/convenient-two-factor-authentication-with-microsoft-passport-and-windows-hello/).

> 1. The keychain can be used with or without biometrics.

I wouldn't constrain it to biometrics but rather if user authentication is required or not. The keychain implementation should then decide what kind of authentication will present to the user.
The reason for this is that on Windows you have no control over what kind of authentication user takes via Windows Hello; be it biometrics, face recognition or PIN. Microsoft Passport API only gives you an option to require user authentication or not.

>2. Ability to use custom encryption for regular QuickUnlock (derived encryption key that isn't saved).

Shall the QuickUnlock feature really be mandatory in case where there is an option to store the key persistently on the system? 🤔 
I kind of feel that you'll most of the time store the key on the system, and won't bother to enter the master key every time for quick unlock option.

> (2) can also be handled with a flag (encrypt/don't) ... 

I'd deffer this to the implementation and not introduce any special flag.

> With fingerprints, the internal encryption key can be derived from the fingerprint data (and not saved), but that's also an implementation detail that doesn't affect the interface.

I don't think that's possible. The biometric authentication usually works on the basis of probability. The received biometric data is not deterministic, thus should not be used for the encryption key material.

> ```c++
> class OSKeychain
> {
>    static bool isAvailable();
>    static bool isPersistent();
>    static const QString& name();
>    bool clear();
>    bool contains(QStringView path) const;
>    bool delete(QStringView path);
>    QSharedPointer<QByteArray> get(QStringView path) const;
>    bool set(QStringView path, const QByteArray& value, const OSKeychain::Flags flags);
> };
> ```

The `delete` is C++ keyword, maybe `remove`?
On Windows, there will also have to be an option to pass pointer to the current `QWidget` for the WindowsHello dialog window to be pinned to it and any message boxes that are to be shown by the implementation. Maybe a constructor with `QWidget*` argument? Or in case `OSKeychain` will be designed in singleton fashion, the `get`/`set` methods should probably have this parameter.

> The flags should include at least OSKeychain::USE_BIOMETRICS, and maybe OSKeychain::ENCRYPT

I'd define only OSKeychain::RequireAuthn (if this is even necessary) and OSKeychain::UsePersistentStorage. Both should be set by default when storing secret key, in my opinion.

**Comment by michaelk83** [comment]
> Windows Hello can't decouple encryption from user authentication (e.g. bio-metrics).

As I understood, you create the encryption key with Microsoft Passport, use that to encrypt the DB passphrase, then store the encrypted passphrase with the Credential API. My impression was that Microsoft Passport is the part that depends on biometrics, but Credential API doesn't. Is that correct?

If that is the case, then when biometrics are disabled, you just need to skip the Microsoft Passport part (specifically, the [`encrypt()` call](https://github.com/keepassxreboot/keepassxc/pull/6029/files#diff-f92bce81551f9b5afb520a30d31933d91db769cccb916357718201220f8897c6R304) in `WinHelloKeyManager::storeKey()` and the [`promptToDecrypt()` call](https://github.com/keepassxreboot/keepassxc/pull/6029/files#diff-f92bce81551f9b5afb520a30d31933d91db769cccb916357718201220f8897c6R332) in `WinHelloKeyManager::getKey()`). Either leave encryption to the caller, or you can do it similar to the [TouchID code](https://github.com/keepassxreboot/keepassxc/blob/develop/src/touchid/TouchID.mm#L59) (that can be pulled out from the TouchID class to a common helper method, maybe in `OSKeychain`). Leaving it to the caller is good enough for the (non-biometric) QuickUnlock use case.

If Credential API also depends on biometrics, then regular QuickUnlock will have to defer to some other key store (that may be a separate implementation). But the class interface can remain the same.

Either way, it's not something you have to worry about in #6029. You can just return "Not supported" for now, and leave that part to #488.

> I wouldn't constrain it to biometrics but rather if user authentication is required or not. ... on Windows you have no control over what kind of authentication user takes via Windows Hello; be it biometrics, face recognition or PIN.

With regular QuickUnlock there will still be a sort of authentication (last few chars of the passphrase, typically), but it'll be handled by KPXC. So maybe call this "OS Authentication"?

> Shall the QuickUnlock feature really be mandatory in case where there is an option to store the key persistently on the system?

The choice is between biometric QuickUnlock, or one based on user input via KPXC (last few chars of passphrase, or maybe a PIN). Users with a fingerprint reader may opt for the former; users without, for the latter.

With Windows Hello there's a "hybrid" option of handling the PIN in the OS. I'm not sure re MacOS, but on Linux I expect the fingerprint support will be just that. There's not going to be OS-level PIN entry.

The common part between these two features is storing the encrypted passphrase in the OS keychain, and using that to unlock the DB without entering the full passphrase. The difference is how the encryption is performed, and whether biometrics (or OS authentication) are involved. So there should be a way to reuse just the storage logic.

It's not a question of "mandatory". Just different code paths that both reuse the same storage logic.

> I'd deffer this to the implementation and not introduce any special [encrypt/don't] flag.

As noted, any internal encryption is transparent to the caller, so I agree that an `ENCRYPT` flag isn't strictly necessary. But OTOH, regular QuickUnlock will do its own encryption in the calling code, and then the internal encryption isn't needed. So when the flag is omitted, it's more of a hint: "I'll handle encryption, so you (the keychain) can skip it".

> The `delete` is C++ keyword, maybe `remove`?

Wasn't sure if `delete` is allowed as a method name in class scope. It's been a while since my C++ days. `remove` is fine.

> pointer to the current `QWidget`

Either of your solutions is fine by me. Another option is a separate `setWidget()` method. Note that the TouchID implementation maintains some global internal state, so it's designed as a singleton. (It may be possible to refactor it to a separate instance per DB, but that's a separate discussion for another time. No need to mess with that now.)

> I'd define only OSKeychain::RequireAuthn (if this is even necessary) and OSKeychain::UsePersistentStorage.

As noted above, consider changing `RequireAuthn` to sometging like `OSAuth`. Authn is specific to Windows, I think. This flag will be needed for the regular QuickUnlock feature, since it will need to handle encryption and user "authentication" outside of the keychain classes (for the case when a fingerprint reader is not available, or the user decides not to use it).

~~Another possible flag is `Overwrite` (or maybe `KeepExisting`) to overwrite or keep already stored keys. I saw you had a boolean parameter for that in your code. But frankly, this can be handled by the caller code using the `contains()` method.~~

**Comment by michaelk83** [comment]
Edited my previous comment to clarify/stress a few bits (you can check the diff).

The more I think about it, the more I tend to agree that the "encrypt" flag isn't needed. In the use cases discussed here it's tied to the OS Authentication flag. They will either be both set, or both cleared.

Your WinHello implementation will probably need to store the OS Auth flag with the passphrase, so that `get()` will know how to handle it (prompt or not). TouchID [controls that with access control flags](https://github.com/keepassxreboot/keepassxc/blob/develop/src/touchid/TouchID.mm#L97) when saving the key.

Looking at your [`WinHelloKeyManager::isAvailable()`](https://github.com/keepassxreboot/keepassxc/pull/6029/files#diff-f92bce81551f9b5afb520a30d31933d91db769cccb916357718201220f8897c6R369), it looks like it may need those flags too, because the answer depends on what the caller needs. Assuming we go with `OSAuth` and `Persistent` flags:
- `isAvailable()`: Is the basic key store available? (Should usually return `true`.)
- `isAvailable(OSAuth)`: Is OS authentication (+ the key store) available?
- `isAvailable(Persistent)`: Is persistent storage available? (With this, `isPersistent()` is no longer needed.)

This leads me to favor an instance (non-singleton) design, where you pass the feature flags (and widget handle) once to the constructor, and then all the methods act accordingly. But this will require a refactor of the TouchID implementation, which I'd prefer to avoid for now (it can be refactored later).

How about this instead:

---

Alternative solution:

1. Split `WinHelloKeyManager` into two layers, with the same class interface. The flags won't be needed with this approach, so they can be dropped:
    1. `WinCredentialStorage`: Just the bare Credential API storage. Persistent storage is implied, and encryption is left to the caller.
    2. `WinHelloKeyManager` (or `WinHelloStorage`, or however you want to call it): Implements encryption via Microsoft Passport, uses `WinCredentialStorage` as the storage backend for the encrypted passphrase. Persistent storage of the encryption key probably also implied?
2. `TouchID` can be similarly split in #488 (at the same opportunity, it may also be changed to non-singleton design):
    1. A base class that just stores the passphrase in the key chain (encryption left to the caller).
    2. A child class that adds encryption and adjusts the access control to require biometrics.
3. Regular QuickUnlock will use the bare storage implementations and handle encryption and authentication on its own.

With that, `isPersistent()` can probably be dropped too, since it's a property of each implementation. The caller can pick whichever it needs. So then the interface boils down to this:
```c++
class OSKeychain
{
    OSKeychain(const QWidget& parent);

    static bool isAvailable();
    static const QString& name();

    bool clear();
    bool contains(QStringView path) const;
    QSharedPointer<QByteArray> get(QStringView path) const;
    bool remove(QStringView path);
    bool set(QStringView path, const QByteArray& value);
};
```
edit: Since the features here are fixed per implementation, this can also be done as a singleton, if preferred.

The encryption and authentication for regular QuickUnlock can be implemented in another class with the same interface, which will use the appropriate OS-specific storage backend to save the encrypted passphrase. Then the difference between all of these variation is just which implementation(s) to use in each case. So we'll have:
- Secure storage layer: `WinCredentialStorage` / `AppleKeychain` / `LinuxKernelKeyring`.
- Encryption & auth layer, biometric:
    - `WinHelloKeyManager` &rarr; uses `WinCredentialStorage` for storage.
    - `TouchID` &rarr; uses `AppleKeychain` for storage.
    - `LinuxFPrint` &rarr; uses `LinuxKernelKeyring` for storage.
- Encryption & auth layer, non-biometric:
    - `QuickUnlock` &rarr; uses either one of the above storage backends, depending on the current OS. Uses a derived encryption key from the last few chars of the passphrase.

All of these share the same class interface, so the application layer can pick whichever one it needs, and have the same code operate it.

**Comment by smlu** [comment]
I've put some more thought into this, and the regular quick unlock which is to be encrypted with chars from passphrase shouldn't be kept in OS storage but only in memory where you can limit to some degree the access to the process memory. 
The reason for this I believe, is the access control to the os storage which is not exclusively limited to the program which stored the secret, and can be accessed by any process (speaking for Windows), unlike on mobile platforms. If any process can access the stored secret and the secret is encrypted with low entropy, in theory a malicious program could bruteforce the master key and steal passwords without leaving a trace. You could still limit the access to the stored secret by requiring user authentication but then in my opinion this feature is meaningless.

>Looking at your WinHelloKeyManager::isAvailable(), it looks like it may need those flags too, because the answer depends on what the caller needs. Assuming we go with OSAuth and Persistent flags:
> * isAvailable(): Is the basic key store available? (Should usually return true.)
> * isAvailable(OSAuth): Is OS authentication (+ the key store) available?
> * isAvailable(Persistent): Is persistent storage available? (With this, isPersistent() is no longer needed.)

Maybe there should also be a function to retrieve storage properties as flags. i.e. you could use then either `isAvailable` or `get<somenameFlags>` and compare flag mask against returned flags value.

>```C++
>class OSKeychain
> {
>    OSKeychain(const QWidget& parent);
>
>   static bool isAvailable();
>    static const QString& name();
>
>   bool clear();
>    bool contains(QStringView path) const;
>   QSharedPointer<QByteArray> get(QStringView path) const;
>    bool remove(QStringView path);
>    bool set(QStringView path, const QByteArray& value);
> };

I think there should probably be also a method to retrieve all mapping keys to the stored secrets. This could be useful in cases where user has moved the kdbx file and KPXC could for example offer to the user to pick from the list of stored secrets.
And maybe a method to change mapping key value to the secret without needing to retrieve the secret first. This could probably be achieved through `set` method.

**Comment by droidmonkey** [comment]
IMO we shouldn't be using paths to define databases. That gets you into trouble on nix systems where symlinks and everything else can ruin your day. Instead, you should use a UUID that uniquely (and anonymously) identifies the open database file that the credentials associate with. 

Technically each database widget has a database, even those that are locked. The locked ones are just blank databases that could be given a uuid to identify it with quick unlock settings.

**Comment by smlu** [comment]
Yeah, using uuid should solve the problems with paths. 
The question then remains what to do with stored secrets of databases which are about to be closed, i.e. how to map the secret back to database when it's reopened or should the secret be deleted from the OS storage? 

While testing the WindowsHello feature I had many times closed and re-opened different databases, and I find it very convenient not to need to re-enter the master key every time the database is opened. 
Also, I'm testing now the case where you have multiple devices and every device stores the database master key. The goal is if I loose the master key I can use one of device to change the master key. If the key is to be deleted from the OS storage when the database is closed it could break such sort of model if you loose track of opened databases. 
I'm not saying due to this cases we should go one or the other way, but I think it might be worth to consider a solid solution here.

**Comment by droidmonkey** [comment]
You can also store data in the unencrypted portion of the kdbx header. You could create a custom data key value pair like "KPXC_PUBLIC_UUID" and store the uuid there. We would have to expose  special header reader that would extract the public key values from a locked database (they are only read on a full database unlock at this point) to correlate the public uuid with what's in os storage. This header data is only supported in kdbx 4. 

https://github.com/keepassxreboot/keepassxc/blob/724f691e159f69c16d01ca00908452d2d8d6c5d4/src/core/Database.h#L101

**Comment by michaelk83** [comment]
> the regular quick unlock which is to be encrypted with chars from passphrase shouldn't be kept in OS storage

Either way, the key point was that all the components share the same class interface, as listed [above](https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-863977242) (give or take final tweaks). @droidmonkey et al can decide in #488 whether or not they want to split out and reuse the OS storage logic, or implement a separate backend, etc. The class interface remains the same, so the aplication layer can use whichever implementation it needs, with the same logic.

And the other notable point was that splitting the fingerprint services vertically as I suggested in the "Alternative solution" part of my [last comment](https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-863977242), allows the OS storage backends to be reused (if desired) without introducing any extra flags etc.

> Maybe there should also be a function to retrieve storage properties as flags.

With the alternative solution, I assumed each implementation's properties are static. As long as they're documented, the programmer chooses the right implementation for whatever they want to accomplish, and there's no need to query the properties at runtime. `isAvailable()` should simply check if the service as a whole is available, with the documented properties implied. No flags are needed in this approach.

*IF* the service properties can vary at runtime, and are not controlled by the implementation class, *then* there may indeed be room for runtime querying. But the implementation can just assume a guaranteed minimum set of properties, and make a best effort to provide the rest (and document that). Then `isAvailable()` would check if those minimum properties are satisfied.

> (additional methods...)

I'll leave these final tweaks to your discretion, but keep in mind that this class interface and all off its implementations will be part of KPXC code. So it'll be fairly easy to add more functions when they are needed.

**Comment by smlu** [comment]
> 
> 
> You can also store data in the unencrypted portion of the kdbx header. You could create a custom data key value pair like "KPXC_PUBLIC_UUID" and store the uuid there. We would have to expose special header reader that would extract the public key values from a locked database (they are only read on a full database unlock at this point) to correlate the public uuid with what's in os storage. This header data is only supported in kdbx 4.
> 
> https://github.com/keepassxreboot/keepassxc/blob/724f691e159f69c16d01ca00908452d2d8d6c5d4/src/core/Database.h#L101

Great idea! I completely missed the custom public data field in the kdbx 4 header. I guess that would be good solution to put file unique id there (maybe KPXC should do this by default anyways in the future 🤔). Looking at the code, I see that the header is read when database is being unlocked. I guess there should be a reader made that parses header in advance when file is being opened but not unlocked. e.g.: `db->open()` reads the header and `db->unlock(key)` decrypts the database.
The other db file formats (other than kdbx 4) I guess we could still use their file paths for storage mapping key, or we skip them altogether as they should probably be upgraded to kdbx 4 anyway (except op vault).

>
>
> Either way, the key point was that all the components share the same class interface,

I totally agree with abstracting this as much as possible. I just think the interface should also take care of restricting access to the stored secret and not leaving any room for permanently storing the secret in an unencrypted (or weakly encrypted) form on the system. As I said before, maybe only for the case when secret is temporary locked in memory, but I think in this case any \*os\* fix should be dropped from the interface name (e.g. Keychain, KeyStore ...).
I think better solution would be to move away from password based quick unlock features, and rather look for any alternative way that the specific platform can provide when biometrics (WindowsHello) is not available on the system. That is, to ensure the persistently stored master key can't be accessed without user needing to authenticate in some form. On Windows this could probably be achieved with DAPI, and I'm certain macOS can also provide some sort of restriction before app can access the stored master key in the Keychain (e.g. user required to enter it's account password) .

**Comment by michaelk83** [comment]
> I just think the interface should also take care of restricting access to the stored secret and not leaving any room for permanently storing the secret in an unencrypted (or weakly encrypted) form on the system.

I think it's ok for the storage backends to leave that responsibility to caller, providing the limitations are properly documented (and the class is named appropriately: the name is part of the documentation). The same way that `File.Save()` isn't responsible for properly encrypting the DB. It's up to the caller to ensure that either A) they use a secure enough implementation, or B) they provide a strong enough encryption (and whatever other protections) on the caller side.

> I think better solution would be to move away from password based quick unlock features, and rather look for any alternative way that the specific platform can provide when biometrics (WindowsHello) is not available on the system.

Maybe so. I agree that storing a weakly encrypted passphrase where it can be accessed by arbitrary processes, is not a good idea. But that can be left to #488. For now, you can leave the credential storage layer inside `WinHelloKeyManager`. If the devs decide to reuse it for QuickUnlock after all, it can be split out later.

**Comment by smlu** [comment]
Does anyone has any idea how secrets could be stored on linux in sys authenticated and/or encrypted way? i.e. fallback mode if biometrics are not supported on the system.

If that's possible then the interface could support only authenticated/encrypted storage by default, and when/if #488 development progresses a special flag could be introduced at that time, e.g. `NoAuthn` which would make the os key storage not to require user authentication.

**Comment by smlu** [comment]
Also I'm re-thinking the need for the `name` method. I think It should be dropped from the interface and instead add special type flags. This type flags then the KEPXC upper layer could use to determine what type of os storage is available. The reason for this is that there can be many storage options at one time (e.g. biometrics fingerprint reader, windowhello, touch id, os authentication/no authentication ...). This can pose an issue to meaningful express those options in a single string (especially for translation).

**Comment by michaelk83** [comment]
> Does anyone has any idea how secrets could be stored on linux in sys authenticated and/or encrypted way?

https://man7.org/linux/man-pages/man7/keyrings.7.html
https://man7.org/linux/man-pages/man7/persistent-keyring.7.html

I haven't read all of it, so I don't know all the limitations and details. Like, is there pid-level access control, etc. But that would be my first go-to. Not sure how that integrates with fprint, either.

Other than that, I've read there was some work to integrate fingerprint login on Linux (probably via fprint). But that will likely depend on the desktop environment and display (login) manager, at least initially.

> when/if #488 development progresses a special flag could be introduced at that time, e.g. `NoAuthn` which would make the os key storage not to require user authentication.

Yes, the main two options to reuse the storage layer are either split it out to its own class (with the same interface, and properly documenting the limitations), or add a flag. That can be decided later. The 3rd option is to *not* reuse the storage layer, but e.g. use KPXC's process memory.

**Comment by maz-1** [comment]
As for windows support, keepassxc depends on libraries that still not available under msvc (libgcrypt) while windows hello support requires WinRT api support, so we really need to wait WinRT support within mingw
https://github.com/msys2/MINGW-packages/issues/6330
https://sourceforge.net/p/mingw-w64/feature-requests/98/

**Comment by QuAzI** [comment]
https://github.com/ShiftMediaProject/libgcrypt ?

**Comment by droidmonkey** [comment]
> As for windows support, keepassxc depends on libraries that still not available under msvc (libgcrypt) while windows hello support requires WinRT api support, so we really need to wait WinRT support within mingw

We moved to Botan and have a PR in the works to fully support msvc compiler.

**Comment by smlu** [comment]
> 
> 
>windows hello support requires WinRT api support...

Actually there is no need for WinRT. The Win32 API is sufficient, and can be used for WindowsHello key storage.
e.g.: #6029 is implemented that way, using only Win32 API.

**Comment by gniusj** [comment]
Abwesenheitsnotiz: 


Vom 08.07.2021  bis einschl. 25.07.2021  befinde ich mich im Urlaub.


E-Mails werden nur verzögert gelesen.
In sehr dringenden Fällen und in Notfällen sprechen Sie bitte eine Nachricht auf den Notruf-Anrufbeantworter unter 02206-95100-0. Geben Sie dabei Name, Rufnummer und die Art des Notfalls an. Wenn es sich um einen sehr dringenden Fall oder einen Notfall handelt, werde ich mich zeitnah bei Ihnen melden.
Andernfalls stehe ich ab dem 26.07.2021 wieder wie gewohnt für Sie zur Verfügung.


Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während ich abwesend bin.


Beste Grüße
Jens Suing

-------------------------------------------------------------------------
Visitenkarte (vcf)
-------------------------------------------------------------------------
DIE Spezialisten - J.Suing
Laurentiusstr. 41 - 51789 Lindlar
fon: 02206-95100-0 - fax: 02206-95100-29
------------------------------------------------------------------------- 
Tobit Software Authorized Partner *****
Starface Advanced Partner 
work ... for all! Solution Partner
bintec elmeg Certified Partner
Kerio Partner 

News, Forum, FAQ und mehr...
www.schneller-und-besser.de 
-------------------------------------------------------------------------
> 
> 
windows hello support requires WinRT api support...

Actually there is no need for WinRT. The Win32 API is sufficient, and can be used for WindowsHello key storage.
e.g.: #6029 is implemented that way, using only Win32 API.

**Comment by QuAzI** [comment]
Still no changes?

**Comment by WinkelCode** [comment]
@QuAzI Stuff seems to be happening in #6029 as of 6 days ago.

I am also very interested in fingerprint reader support, already using it on my iPad via a different app and it's very convenient there.

**Comment by QuAzI** [comment]
Damn. Just found that [IssueHunt is abandoned](https://github.com/keepassxreboot/keepassxc/issues/5991#issuecomment-881954197) and it is pity. Maybe better to remove links to this platform to prevent donations lose?

Is there any plans to add this feature into [milestone 2.7](https://github.com/keepassxreboot/keepassxc/milestone/27) like #5991? @droidmonkey ?

**Comment by droidmonkey** [comment]
I have no way to actually test an implementation of this. I don't think this would make it in 2.7

**Comment by droidmonkey** [comment]
Windows Hello integration will be part of 2.7, so yes well have fingerprint support in Windows!

**Comment by QuAzI** [comment]
Waiting for him more than a gift for the new year

**Comment by gniusj** [comment]
Abwesenheitsnotiz: 


Vom 23.12.2021  bis einschl. 09.01.2022  befinde ich mich im Urlaub.


E-Mails werden nur verzögert gelesen.
In sehr dringenden Fällen und in Notfällen sprechen Sie bitte eine Nachricht auf den Notruf-Anrufbeantworter unter 02206-95100-0. Geben Sie dabei Name, Rufnummer und die Art des Notfalls an. Wenn es sich um einen sehr dringenden Fall oder einen Notfall handelt, werde ich mich zeitnah bei Ihnen melden.
Andernfalls stehe ich ab dem 10.01.2022 wieder wie gewohnt für Sie zur Verfügung.


Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während ich abwesend bin.


Beste Grüße
Jens Suing

-------------------------------------------------------------------------
Visitenkarte (vcf)
-------------------------------------------------------------------------
DIE Spezialisten - J.Suing
Laurentiusstr. 41 - 51789 Lindlar
fon: 02206-95100-0 - fax: 02206-95100-29
------------------------------------------------------------------------- 
Tobit Software Authorized Partner *****
Starface Advanced Partner 
work ... for all! Solution Partner
bintec elmeg Certified Partner
Kerio Partner 

News, Forum, FAQ und mehr...
www.schneller-und-besser.de 
-------------------------------------------------------------------------
Waiting for him more than a gift for the new year

**Comment by QuAzI** [comment]
Released! But...
Now if fingerprint scanner present KeePassXC forcing to use it.
If 'cancel' pressed I can't unlock DB anymore while password not entered.
Fingerprint ignores Key File. So we lose one protection factor.

**Comment by goetzc** [comment]
I think that’s the case because finger biometric  unlock is “quick unlock” and not full unlock.

**Comment by tio-trom** [comment]
So no Linux support for the fingerprint reader?

**Comment by GonzRon** [comment]
"brutePrint" sheds some light on fingerprint cracking - not very secure on android, slightly better on iOS due to encrypted chip level communication.
https://www.kaspersky.com/blog/fingerprint-brute-force-android/48303/

**Comment by anonymous-user** [comment]
What a shame a Google Pixel wasn't included. I would have liked to know how it'd fare.
